db/background_mgr_attachment_migration.go

```diff
@@ -0,0 +1,351 @@
+			return true
+		}
+
+		a.DocsProcessed.Add(1)
+		syncData, _, _, err := UnmarshalDocumentSyncDataFromFeed(event.Value, event.DataType, collection.userXattrKey(), false)
+		if err != nil {
+			failProcess(err, "error unmarshaling document %s: %v", base.UD(docID), err)
+		}
+
+		if syncData == nil || syncData.Attachments == nil {
+			// no attachments to migrate
+			return true
+		}
+
+		collCtx := collection.AddCollectionContext(ctx)
+		collWithUser := &DatabaseCollectionWithUser{
+			DatabaseCollection: collection,
+		}
+		// xattr migration to take place
+		err = collWithUser.MigrateAttachmentMetadata(collCtx, docID, event.Cas, syncData)
+		if err != nil {
+			failProcess(err, "error migrating document attachment metadata for doc: %s: %v", base.UD(docID), err)
+		} else {
+			a.DocsChanged.Add(1)
+		}
+		return true
+	}
+
+	bucket, err := base.AsGocbV2Bucket(db.Bucket)
+	if err != nil {
+		return err
+	}
+
+	currCollectionIDs := db.GetCollectionIDs()
+	checkpointPrefix := db.MetadataKeys.DCPCheckpointPrefix(db.Options.GroupID) + "att_migration:"
+
+	// check for mismatch in collection id's between current collections on the db and prev run
+	err = a.resetDCPMetadataIfNeeded(ctx, db.MetadataStore, db, checkpointPrefix, currCollectionIDs)
+	if err != nil {
+		return err
+	}
+
+	a.SetCollectionIDs(currCollectionIDs)
+	dcpOptions := getMigrationDCPClientOptions(currCollectionIDs, db.Options.GroupID, checkpointPrefix)
+	dcpFeedKey := GenerateAttachmentMigrationDCPStreamName(a.MigrationID)
+	dcpClient, err := base.NewDCPClient(ctx, dcpFeedKey, callback, *dcpOptions, bucket)
+	if err != nil {
+		base.WarnfCtx(ctx, "[%s] Failed to create attachment migration DCP client: %v", migrationLoggingID, err)
+		return err
+	}
+	base.InfofCtx(ctx, base.KeyAll, "[%s] Starting DCP feed %q for attachment migration", migrationLoggingID, dcpFeedKey)
+
+	doneChan, err := dcpClient.Start()
+	if err != nil {
+		base.WarnfCtx(ctx, "[%s] Failed to start attachment migration DCP feed: %v", migrationLoggingID, err)
+		_ = dcpClient.Close()
+		return err
+	}
+	base.DebugfCtx(ctx, base.KeyAll, "[%s] DCP client started for Attachment Migration.", migrationLoggingID)
+
+	select {
+	case <-doneChan:
+		base.InfofCtx(ctx, base.KeyAll, "[%s] Finished migrating attachment metadata from sync data to global sync data. %d/%d docs changed", migrationLoggingID, a.DocsChanged.Value(), a.DocsProcessed.Value())
+		err = dcpClient.Close()
+		if err != nil {
+			base.WarnfCtx(ctx, "[%s] Failed to close attachment migration DCP client! %v", migrationLoggingID, err)
+			return err
+		}
+		if processFailure != nil {
+			return processFailure
+		}
+		// set sync info here
+		for _, collectionID := range currCollectionIDs {
+			dbc, ok := db.CollectionByID[collectionID]
+			if !ok {
+				base.InfofCtx(ctx, base.KeyAll, "[%s] Completed attachment migration, but unable to update syncInfo for collection %v (collection is no longer associated with the database)", migrationLoggingID, collectionID)
```

Comment:
I don't quite see how this can happen, the dcp client is tied to a `DatabaseContext` and will not run without it. You can't change `CollectionByID` while a `DatabaseContext` exists.

It's possible that someone will change the collections while this background job is running, but then that will issue `DatabaseContext.Close` which will close the terminators, and hit `terminator.Done()`. In the very unlikely scenario that the dcp client finishes at the same time, `db.CollectionByID` will still be present.

We definitely don't want to panic in any circumstances, so if this is possible to hit, we should handle it (and write a test). But I think that as is, that error would be not a possibility and you can just do `collection := db.CollectionByID[collectionID]`

Reply:
I will remove, I wanted to be extra safe given this handling is also included in the resync dcp task but will remove.